### PR TITLE
Jetpack Focus: Disable notifications and deep links during phase 4

### LINKED ISSUE #19812: Jetpack Focus: Correctly handle notification and widget taps and deep links in phase 4

### PR DESCRIPTION
**Pull request by hassaanelgarem**
Closes #19812

## Description
This PR disables various entry points to Jetpack features during phase 4. It also completely disables notifications. 

### Normal state

1. Make sure the Jetpack app is not installed on your testing device
2. Make sure you've allowed notifications for the WordPress app
3. Launch the app
4. Open the debug menu and disable all the Jetpack Features Removal flags
5. Kill and relaunch the app
6. Make sure remote notifications are working (You can use the MC tool)
7. Make sure the weekly roundup notification is working (Detailed testing instructions can be found [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/19364))
8. Copy the following links to a note on your device:
    - wordpress.com/stats
    - wordpress.com/read
    - wordpress.com/notifications
    - wordpress.com/me
9. Tap on each deep link, and make sure it opens the app and navigates to the respective feature
10. Send the app to the background
11. Long press on the app's icon
12. Make sure that the displayed menu has items for stats and notifications

### Phase 4 state

1. Make sure the Jetpack app is not installed on your testing device
2. Make sure you've allowed notifications for the WordPress app
3. Launch the app
4. Open the debug menu and enable "Jetpack Features Removal Phase Four"
5. Kill and relaunch the app
6. Make sure remote notifications are disabled (You can use the MC tool)
7. Make sure the weekly roundup notification is disabled (Detailed testing instructions can be found [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/19364))
8. Copy the following links to a note on your device:
    - wordpress.com/stats
    - wordpress.com/read
    - wordpress.com/notifications
    - wordpress.com/me
9. For the first three links
    1. Tap on the link
    2. The app is opened
    3. Make sure the corresponding feature is not displayed
    4. Check the logs for a `jetpack_feature_incorrectly_accessed` tracked event with `{"calling_function": "deep_link"}` 
10. For the /me link, tap on it and make sure it opens the app and displays the Me scene. 
11. Send the app to the background
12. Long press on the app's icon
13. Make sure that the displayed menu does not have items for stats and notifications

## Regression Notes
1. Potential unintended areas of impact
N/A

5. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

6. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Utility/PushNotificationsManager.swift [hunk 1]
```diff
@@ -53,18 +53,31 @@ final public class PushNotificationsManager: NSObject {
...
         sharedApplication.registerForRemoteNotifications()
+        didRegisterForRemoteNotifications = true
+        return
     }
 
-
+    private func disableRemoteNotifications() {
+        if !didRegisterForRemoteNotifications {
+            sharedApplication.registerForRemoteNotifications()
+        }
```
Comment (hassaanelgarem) [inline WordPress/Classes/Utility/PushNotificationsManager.swift:76]: Turns out, calling `unregisterForRemoteNotifications()` while `registerForRemoteNotifications()` has not been called during this launch does nothing.

Without this, disabling notifications on launch doesn't properly work.

**Review by staskus — COMMENTED** [review]
The changes look good 👍

I tested given scenarios as well as done additional notification migration testing, to make sure it stays unaffected. The `notificationsDisabled || appMigrated` logic is a correct one to avoid any issues 👍.

I found one small corner-case'y issue though, notifications are not disabled, if the application is launched from a notification:

1. Enable WordPress app notifications
2. Enable Phase 4 
3. Kill the app
4. Send a remote notification
5. Click on a notification
6. WordPress app is launched
7. Send a remote notification
8. Notification is still received

I suspect notification disabling logic needs to be called in one more place where remote notification logic is handled.

**Comment by hassaanelgarem** [comment]
> notifications are not disabled, if the application is launched from a notification:

Nice catch @staskus!
This should be fixed by 50ecc623cefd2881b926903b6bab085865b65bc9 👍

**Comment by staskus** [comment]
> > notifications are not disabled, if the application is launched from a notification:
> 
> Nice catch @staskus! This should be fixed by [50ecc62](https://github.com/wordpress-mobile/WordPress-iOS/commit/50ecc623cefd2881b926903b6bab085865b65bc9) 👍

Thanks for the fix!

I still continue receiving notifications after opening the app from notification, so the fix might not be enough, @hassaanelgarem .

**Comment by hassaanelgarem** [comment]
@staskus That's weird! It's working well for me 🤔 

I'll give this another look. However, given that this is an edge case, I think we should not block this PR by it. Instead, I've created #19840 to track this. Let me know if you agree.

**Comment by staskus** [comment]
> @staskus That's weird! It's working well for me 🤔
> 
> I'll give this another look. However, given that this is an edge case, I think we should not block this PR by it. Instead, I've created #19840 to track this. Let me know if you agree.

Okay, sounds fair 👍

**Review by staskus — APPROVED** [review]

**Pull request merged**